hugr-py/src/hugr/hugr/render.py

```diff
@@ -0,0 +1,280 @@
+        <TABLE BORDER="0" CELLBORDER="0">
+            <TR><TD><FONT POINT-SIZE="{fontsize}" FACE="{fontface}"
+                COLOR="{label_color}"><B>{node_label}</B>{node_data}</FONT></TD></TR>
+        </TABLE>
+        </TD>
+    </TR>
+    {outputs_row}
+    </TABLE>
+    """
+
+    _HTML_PORTS_ROW_TEMPLATE = """
+        <TR>
+            <TD>
+                <TABLE BORDER="0" CELLBORDER="0" CELLSPACING="3" CELLPADDING="2">
+                    <TR>
+                        {port_cells}
+                    </TR>
+                </TABLE>
+            </TD>
+        </TR>
+    """
+
+    _HTML_PORT_TEMPLATE = (
+        '<TD BGCOLOR="{back_colour}" COLOR="{border_colour}"'
+        ' PORT="{port_id}" BORDER="{border_width}">'
+        '<FONT POINT-SIZE="10.0" FACE="{fontface}" COLOR="{font_colour}">{port}</FONT>'
+        "</TD>"
+    )
+
+    _INPUT_PREFIX = "in."
+    _OUTPUT_PREFIX = "out."
+
+    def _format_html_label(self, **kwargs: str) -> str:
+        _HTML_LABEL_DEFAULTS = {
+            "label_color": self.palette.dark,
+            "node_back_color": self.palette.node,
+            "inputs_row": "",
+            "outputs_row": "",
+            "border_colour": self.palette.port_border,
+            "border_width": "1",
+            "fontface": self._FONTFACE,
+            "fontsize": 11.0,
+        }
+        return self._HTML_LABEL_TEMPLATE.format(**{**_HTML_LABEL_DEFAULTS, **kwargs})
+
+    def _html_ports(self, ports: Iterable[str], id_prefix: str) -> str:
+        return self._HTML_PORTS_ROW_TEMPLATE.format(
+            port_cells="".join(
+                self._HTML_PORT_TEMPLATE.format(
+                    port=port,
+                    # differentiate input and output node identifiers
+                    # with a prefix
+                    port_id=id_prefix + port,
+                    back_colour=self.palette.background,
+                    font_colour=self.palette.dark,
+                    border_width="1",
+                    border_colour=self.palette.port_border,
+                    fontface=self._FONTFACE,
+                )
+                for port in ports
+            )
+        )
+
+    def _in_port_name(self, p: InPort) -> str:
+        return f"{p.node.idx}:{self._INPUT_PREFIX}{p.offset}"
+
+    def _out_port_name(self, p: OutPort) -> str:
+        return f"{p.node.idx}:{self._OUTPUT_PREFIX}{p.offset}"
+
+    def _in_order_name(self, n: Node) -> str:
+        return f"{n.idx}:{self._INPUT_PREFIX}None"
+
+    def _out_order_name(self, n: Node) -> str:
+        return f"{n.idx}:{self._OUTPUT_PREFIX}None"
+
+    def _viz_node(self, node: Node, hugr: Hugr, graph: Digraph) -> None:
+        """Render a (possibly nested) node to a graphviz graph."""
+        # TODO: Port the CFG special-case rendering from guppy, and use it here
+        # when a node is a CFG node.
+        # See https://github.com/CQCL/guppylang/blob/7d5106cd59ad452046d0dfffd10eea9d9b617431/guppylang/hugr_builder/visualise.py#L250
```

Comment:
This is purely for debugging Guppy's dataflow analysis, so probably doesn't belong here